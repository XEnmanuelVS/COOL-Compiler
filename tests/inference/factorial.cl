class Main {
    fact(x: AUTO_TYPE): AUTO_TYPE {
        if x = 0 then 1 else x * fact(x - 1) fi
    };
};
