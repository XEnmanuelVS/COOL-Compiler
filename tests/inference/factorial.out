\__ProgramNode [<class> ... <class>]
  \__ClassDeclarationNode: class Main  { <feature> ... <feature> }
    \__FuncDeclarationNode: fact(x: Int) : Int -> <body>
      \__ConditionalNode: if <expr> then <expr> else <expr> fi
        \__if 
          \__<expr> EqualNode <expr>
            \__VariableNode: x
            \__IntegerNode: 0
        \__then 
          \__IntegerNode: 1
        \__else 
          \__<expr> StarNode <expr>
            \__VariableNode: x
            \__CallNode: <obj>.fact(<expr>, ..., <expr>)
              \__VariableNode: self
              \__<expr> MinusNode <expr>
                \__VariableNode: x
                \__IntegerNode: 1

